### PR TITLE
Cordon node

### PR DESCRIPTION
**Pull request by suiguoxin**
Cordon node when NvidiaSmiEccError: 
- collect node_name in job_exporter
- set alert in Prometheus
- add `cordon-nodes` action in alert-manager, cordon nodes through k8s api `patchNode`

Refer to Issue #4789

### CODE REVIEW

**Comment by fanyangCS** [comment]
Maybe you could also implement  uncordon as well?

**Comment by Binyang2014** [comment]
> Maybe you could also implement uncordon as well?

I suggest uncordon node manually. I think alert resolve message may be false positive. Such as job-exporter trigger the alert, then the job exporter is crash.  Since there is no metric anymore, the alert rules is not satisfied and the alert maybe resolved by Prometheus. But error still here.

So I suggest admin uncordon the node manually when he make sure everything is OK

File: src/alert-manager/deploy/rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  name: alert-manager-account
+  namespace: default
+---
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: alert-manager-role
+rules:
+  - apiGroups: [""]
+    resources: ["nodes", "pods", "persistentvolumes", "persistentvolumeclaims"]
```
Comment (Binyang2014) [inline src/alert-manager/deploy/rbac.yaml:16]: Only need nodes permission?

File: src/alert-manager/deploy/stop.sh [hunk 2]
```diff
@@ -19,3 +19,11 @@
...
 kubectl delete --ignore-not-found --now configmap/alertmanager
 kubectl delete --ignore-not-found --now deployment/alertmanager
+
+if kubectl get clusterrolebinding | grep -q "alert-manager-role-binding"; then
+    kubectl delete clusterrolebinding alert-manager-role-binding || exit $?
+fi
+
+if kubectl get serviceaccount | grep -q "alert-manager-account"; then
+    kubectl delete serviceaccount alert-manager-account || exit $?
+fi
```
Comment (Binyang2014) [inline src/alert-manager/deploy/stop.sh:29]: Delete `ClusterRole` here?

File: src/alert-manager/src/alert-handler/controllers/node.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      const node = alert.labels.host_name;
+      // set the node unschedulable
+      k8sApi
+        .patchNode(node, { spec: { unschedulable: true } })
+        .then(function () {
+          logger.info(`alert-handler successfully cordon node ${node}`);
+        })
+        .catch(function (data) {
+          logger.error(`alert-handler failed to cordon node ${node}`);
+          logger.error(data);
```
Comment (Binyang2014) [inline src/alert-manager/src/alert-handler/controllers/node.js:44]: Merge the two logs? 
```js
 logger.error(`alert-handler failed to cordon node ${node}, resp data: ${data}`);
```

File: src/alert-manager/src/alert-handler/controllers/node.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+kc.loadFromDefault();
+const k8sApi = kc.makeApiClient(k8s.CoreV1Api);
+
+const cordonNodes = (req, res) => {
+  logger.info(
+    'alert-handler received `cordonNode` post request from alert-manager.',
+  );
+
+  logger.info(req.body);
```
Comment (Binyang2014) [inline src/alert-manager/src/alert-handler/controllers/node.js:30]: Merge the two log lines?

File: src/job-exporter/deploy/job-exporter.yaml.template [hunk 5]
```diff
@@ -65,6 +65,10 @@ spec:
           valueFrom:
             fieldRef:
               fieldPath: status.hostIP
+        - name: HOST_NAME
```
Comment (Binyang2014) [inline src/job-exporter/deploy/job-exporter.yaml.template:68]: Using NODE_NAME here? Same as k8s node name

File: src/job-exporter/src/collector.py [hunk 6]
```diff
@@ -414,8 +414,8 @@ def convert_nvidia_gpu_info_to_metrics(gpu_info, zombie_info, pid_to_cid_fn, mem
             nvidia_mem_utils.add_metric([minor], info.gpu_mem_util)
             if info.temperature is not None:
                 nvidia_gpu_temp.add_metric([minor], info.temperature)
-            nvidia_ecc_errors.add_metric([minor, "single"], info.ecc_errors.single)
-            nvidia_ecc_errors.add_metric([minor, "double"], info.ecc_errors.double)
+            nvidia_ecc_errors.add_metric([host_name, minor, "single"], info.ecc_errors.single)
```
Comment (Binyang2014) [inline src/job-exporter/src/collector.py:417]: Change to `node_name`?

File: src/alert-manager/deploy/stop.sh (hunk 2 above)
Comment (suiguoxin) [reply src/alert-manager/deploy/stop.sh:29]: > `ClusterRole`

Deleted, thanks

File: src/alert-manager/deploy/rbac.yaml (hunk 1 above)
Comment (suiguoxin) [reply src/alert-manager/deploy/rbac.yaml:16]: > Only need nodes permission?

Fixed

File: src/alert-manager/src/alert-handler/controllers/node.js (hunk 3 above)
Comment (suiguoxin) [reply src/alert-manager/src/alert-handler/controllers/node.js:44]: > Merge the two logs?
> 
> ```js
>  logger.error(`alert-handler failed to cordon node ${node}, resp data: ${data}`);
> ```

Merged, thanks

File: src/alert-manager/src/alert-handler/controllers/node.js (hunk 4 above)
Comment (suiguoxin) [reply src/alert-manager/src/alert-handler/controllers/node.js:30]: > Merge the two log lines?

Second log line deleted.

File: src/job-exporter/deploy/job-exporter.yaml.template (hunk 5 above)
Comment (suiguoxin) [reply src/job-exporter/deploy/job-exporter.yaml.template]: > Using NODE_NAME here? Same as k8s node name

changed, thanks

File: src/job-exporter/src/collector.py (hunk 6 above)
Comment (suiguoxin) [reply src/job-exporter/src/collector.py]: > Change to `node_name`?

Changed.

**Comment by suiguoxin** [comment]
> > Maybe you could also implement uncordon as well?
> 
> I suggest uncordon node manually. I think alert resolve message may be false positive. Such as job-exporter trigger the alert, then the job exporter is crash. Since there is no metric anymore, the alert rules is not satisfied and the alert maybe resolved by Prometheus. But error still here.
> 
> So I suggest admin uncordon the node manually when he make sure everything is OK

@fanyangCS @Binyang2014 Added a brief instruction for admin about the alert and how to uncordon nodes here. 
https://github.com/suiguoxin/pai/blob/cordon-node/docs/manual/cluster-admin/troubleshooting.md#nvidiasmieccerror

File: src/alert-manager/deploy/alert-manager-configmap.yaml.template [hunk 7]
```diff
@@ -32,8 +32,12 @@ data:
       repeat_interval: {{ cluster_cfg["alert-manager"]["repeat-interval"] }}
       group_by: [alertname, alertstate]
 
-      {% if 'routes' in cluster_cfg["alert-manager"]["customized-routes"] %}
       routes:
+      - receiver: pai-cordon-nodes
+        match:
+          alertname: NvidiaSmiEccError
```
Comment (Binyang2014) [inline src/alert-manager/deploy/alert-manager-configmap.yaml.template:38]: Seems only need to match double EccError here. Please only match the alert when label `type` is `double`

File: docs/manual/cluster-admin/troubleshooting.md [hunk 8]
```diff
@@ -67,6 +67,15 @@ If you cannot use GPU in your job, please check the following items on the corre
 
 If the GPU number shown in webportal is wrong, check the [hivedscheduler and VC configuration](./how-to-set-up-virtual-clusters.md).
 
+### NvidiaSmiEccError
+This is a kind of alert from alert manager.
+It means that nvidia cards from the related nodes have double ecc error.
```
Comment (Binyang2014) [inline docs/manual/cluster-admin/troubleshooting.md:72]: NvidiaSmiEccError will be fired when there is double/single ECC error, we only crodon the node when there is `double` ECC error

**Review by Binyang2014 — APPROVED** [review]

**Pull request merged**